File: Problems/0530_Minimum_Absolute_Difference_in_BST/Minimum_Absolute_Difference_in_BST.cs
using System;
using System.Collections.Generic;

/* Definition for a binary tree node. */
 public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}
 
public class Solution
{
    public int GetMinimumDifference(TreeNode root)
    {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode curr = root;
        int diff = int.MaxValue;
        int prev = 0;
        bool first = true;
        while (curr != null || stack.Count > 0)
        {
            if (curr != null)
            {
                stack.Push(curr);
                curr = curr.left;
            }
            else
            {
                curr = stack.Pop();
                if (!first) diff = Math.Min(diff, Math.Abs(prev - curr.val));
                else first = false;
                prev = curr.val;
                
                curr = curr.right;
            }
        }
        return diff;
    }
    public IList<int> vals;
    public int GetMinimumDifference2(TreeNode root)
    {
        vals = new List<int>();
        get_all_vals2(root);
        int diff_min = int.MaxValue;
        for (int i = 0; i < vals.Count; ++i)
        {
            for (int j = i + 1; j < vals.Count; ++j)
            {
                int temp = Math.Abs(vals[i] - vals[j]);
                if (temp < diff_min)
                    diff_min = temp;
            }
        }
        vals = null;
        return diff_min;
    }

    private void get_all_vals2(TreeNode node)
    {
        if (node == null)
            return;
        vals.Add(node.val);
        if (node.left != null)
            get_all_vals2(node.left);
        if (node.right != null)
            get_all_vals2(node.right);
        return;
    }

    private TreeNode set_node(string[] flds, int depth, int pos)
    {
        if (flds.Length == 0)
            return null;

        int cur_pos = 0;
        for (int i = 0; i < depth; ++i)
            cur_pos += (int)Math.Pow(2, i);
        
        if (cur_pos + pos > flds.Length - 1)
            return null;
        
        if (flds[cur_pos + pos] == "null")
            return null;

        TreeNode node = new TreeNode(int.Parse(flds[cur_pos + pos]));
        node.left = set_node(flds, depth + 1, 2*pos);
        node.right = set_node(flds, depth + 1, 2*pos + 1);

        return node;
    }

    public string output_int_array(int[] nums)
    {
        if (nums.Length <= 0)
            return "";

        string resultStr = "[" +  nums[0].ToString();
 
        for (int i = 1; i < nums.Length; ++i)
        {
            resultStr += "," + nums[i].ToString();
        }

        return resultStr + "]";
    }


    public void Main(string args)
    {
        string flds = args.Replace("[[","").Replace("]]","").Trim();
        TreeNode root = set_node(flds.Split(','), 0, 0);
        Output_TreeNode ot = new Output_TreeNode();

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        int result = GetMinimumDifference(root);
        Console.WriteLine("result = " + result.ToString());
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
    }
}

public class Output_TreeNode
{
    List<string> resultStr = new List<string>();
    public string output(TreeNode node)
    {
        output_tree(node, 0);
        return print_result();
    }
    public void output_tree(TreeNode node, int n)
    {
        if (node == null)
            return;
        
        if (resultStr.Count <= n)
            resultStr.Add("(" + node.val.ToString() + ")");
        else
            resultStr[n] += ",(" + node.val.ToString() + ")";

        if (node.left != null)
            output_tree(node.left, n + 1);
        if (node.right != null)
            output_tree(node.right, n + 1);

        return;
    }

    public string print_result()
    {
        string outputStr = "";

        for (int i = 0; i < resultStr.Count; ++i)
        {
            outputStr += resultStr[i] + "\n";
        }

        return outputStr;
    }
}

File: Problems/0530_Minimum_Absolute_Difference_in_BST/Project_Python3/Minimum_Absolute_Difference_in_BST.py
import os
import sys
import time

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def getMinimumDifference0(self, root):
        L = []
        def dfs(node):
            if node.left: dfs(node.left)
            L.append(node.val)
            if node.right: dfs(node.right)
        dfs(root)
        return min(abs(a - b) for a, b in zip(L, L[1:]))

    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.vals = []
        self.get_all_vals2(root)
        '''
        diff_min = sys.maxsize
        for i in range(len(self.vals)):
            for j in range(i + 1, len(self.vals)):
                temp = abs(self.vals[i] - self.vals[j])
                if temp < diff_min:
                    diff_min = temp
        return diff_min
        '''
        return min(abs(a - b) for a, b in zip(self.vals, self.vals[1:]))

    def get_all_vals2(self, node):
        if node == None:
            return
        self.vals.append(node.val)
        if node.left != None:
            self.get_all_vals2(node.left)
        if node.right != None:
            self.get_all_vals2(node.right)
        return

    def getMinimumDifference3(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.vals_count = 0
        self.get_all_vals3_1(root)
        self.vals = [0]*self.vals_count
        self.vals_count = 0
        self.get_all_vals3_2(root)
        diff_min = sys.maxsize
        for i in range(len(self.vals)):
            for j in range(i + 1, len(self.vals)):
                temp = abs(self.vals[i] - self.vals[j])
                if temp < diff_min:
                    diff_min = temp
        return diff_min

    def get_all_vals3_1(self, node):
        if node == None:
            return
        self.vals_count += 1
        if node.left != None:
            self.get_all_vals3_1(node.left)
        if node.right != None:
            self.get_all_vals3_1(node.right)
        return

    def get_all_vals3_2(self, node):
        if node == None:
            return
        self.vals[self.vals_count] = node.val
        self.vals_count += 1
        if node.left != None:
            self.get_all_vals3_2(node.left)
        if node.right != None:
            self.get_all_vals3_2(node.right)
        return

class output_TreeNode:
    def __init__(self):
        self.resultStr = []

    def output(self, node):
        self.output_TreeNode(node, 0)
        return self.print_resultStr()

    def output_TreeNode(self, node, n):
        if node == None:
            return
        if len(self.resultStr) <= n:
            self.resultStr.append("(" + str(node.val) + ")")
        else:
            self.resultStr[n] += ",(" + str(node.val) + ")"
        if node.left != None:
            self.output_TreeNode(node.left, n + 1)
        if node.right != None:
            self.output_TreeNode(node.right, n + 1)
        return

    def print_resultStr(self):
        outputStr = ""
        for i in range(len(self.resultStr)):
            outputStr += self.resultStr[i] + "\n"
        return outputStr

def set_node(flds, depth, pos):
    if len(flds) <= 0:
        return None

    cur_pos = 0
    for i in range(depth):
        cur_pos += 2 ** i
    
    if cur_pos + pos > len(flds) - 1:
        return None

    if flds[cur_pos + pos] == 'null':
        return None

    node = TreeNode(int(flds[cur_pos + pos]))
    node.left = set_node(flds, depth + 1, 2*pos)
    node.right = set_node(flds, depth + 1, 2*pos + 1)

    return node

def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()

def loop_main(temp):
    flds = temp.replace("\"","").replace("[","").replace("]","").rstrip().split(",")
    root = set_node(flds, 0, 0)

    time0 = time.time()

    sl = Solution()
    result = sl.getMinimumDifference(root)

    print("result = %d" %result)

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()

if __name__ == "__main__":
    main()
